### ISSUE TITLE
Bug: [no-useless-default-assignment] Incorrect warning with React useCallback (`T extends Function`)

### ISSUE THREAD

**Issue by stephenwade** [typescript-eslint/typescript-eslint#11874]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABCAzgUwMIEMA2OBGWEA1gDwAqiaAHlGmACYqIBi408YAfABQS4EixAFyJyAGkQB9BmgAO9WZACeAGRgooorGGUBtALoBKRAG8AUIkQAnNFBDWk-PIRIBucwF9z5mnLjWUIigkLAIiACyyhhwALb%2BYPRQPCYWVhAImogQABZoJORY1gDmdogAvMjo2C5CPDymiDDFYAGYCHS0FcG46Iie3aaeJuVcZpZWNnYOSM2ttjFgnVAeVp6ShkYeE7n5xIUldimr2XkFRaXJjXNti8uiUNYgaP1bXuZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0UAO7joTSODABfEPqA&tsconfig=N4XyA&tokens=false

### Repro Code

```TypeScript
function useCallback<T extends Function>(callback: T, _dependencyList: any[]) {
  return callback;
}

export function MyComponent() {
  const checkTarget = useCallback(({ ignoreContext = false } = {}) => {
    return ignoreContext;
  }, []);

  checkTarget();
  checkTarget({ ignoreContext: true });
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-useless-default-assignment": "warn"
  }
}
```

### tsconfig

```jsonc
{}
```

### Expected Result

I expected that there would be no warnings.

### Actual Result

There is a warning on the `{}`:

```
  6:64  warning  Default value is useless because the parameter is not optional  @typescript-eslint/no-useless-default-assignment
```

If I apply the autofix (removing `= {}`), it causes a TypeScript error on line 10 when calling `checkTarget` with no arguments.

### Additional Info

This seems to be related to the way React defines `useCallback` with a generic type parameter.

I copied the type definition from React to make my example minimal. If I make this change to my `useCallback` definition, the rule doesn't warn:

```diff
-function useCallback<T extends Function>(callback: T, _dependencyList: any[]) {
+function useCallback(callback: Function, _dependencyList: any[]) {
```

**Comment by kirkwaiblinger** [comment]
Hi @stephenwade! This is a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/11846 (specifically the `useCallback` part). 

This has been fixed on main already (see [playground](https://main--typescript-eslint.netlify.app/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABCAzgUwMIEMA2OBGWEA1gDwAqiaAHlGmACYqIBi408YAfABQS4EixAFyJyAGkQB9BmgAO9WZACeAGRgooorGGUBtALoBKRAG8AUIkQAnNFBDWk-PIRIBucwF9z5mnLjWUIigkLAIiACyyhhwALb%2BYPRQPCYWVhAImogQABZoJORY1gDmdogAvMjo2C5CPDymiDDFYAGYCHS0FcG46Iie3aaeJuVcZpZWNnYOSM2ttjFgnVAeVp6ShkYeE7n5xIUldimr2XkFRaXJjXNti8uiUNYgaP1bXkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0UAO7joTSODABfEPqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)) and will be released shortly with [our next scheduled release](https://typescript-eslint.io/users/releases).
